merkle/merkle_path.go

```diff
@@ -152,19 +164,59 @@ func pathFromNodeToRootAtSnapshot(node int64, level int, snapshot int64, maxBitL
 	return proof, nil
 }
 
-// subtreeDepth calculates the depth of a subtree, used at the right of the tree which
-// may not be completely populated
-func subtreeDepth(size int64, bits int) int {
-	for b := bitLen(size) - 1; b > bits; b-- {
-		size = size &^ (1 << uint(b))
+// lastNodeWritten determines if the last node is present in storage for a given Merkle tree size
+// and level in the tree (0 = leaves, increasing towards the root). This is determined by
+// examining the bits of the last valid leaf index in a tree of the specified size. Zero bits
+// indicate nodes that are not stored at that tree size.
+//
+// Example:
+//
+//             hash2
+//             /  \
+//            /    \
+//           /      \
+//          /        \
+//         /          \
+//        k            [X]
+//       / \             \
+//      /   \             \
+//     /     \             \
+//    g       h           [X]
+//   / \     / \         /
+//   a b     c d        e
+//   | |     | |        |
+//   d0 d1   d2 d3      d4
+//
+// Tree size = 5, last index = 4 in binary = 100, append 1 for leaves = 1001.
+// Reading down the RHS: present, not present, not present, present = 1001. So when
+// attempting to fetch the sibling of k the tree should be descended twice to fetch
+// 'e'.
+func lastNodeWritten(d, ts int64) bool {
+	if d == 0 {
+		// Leaves always exist
+		return true
 	}
 
-	// determine tree height for the remaining bits.
-	p2 := bitLen(size) - 1
-	size = size &^ (1 << uint(p2))
-	if bitLen(size) > 0 {
-		p2++
+	// Last index is size - 1, we shift one to avoid special case (leaves always present)
+	bits := uint64(ts - 1)
+	// Test the bit in the path for the requested level
+	mask := uint64(1) << uint64(d-1)
+
+	return bits&mask != 0
+}
+
+// skipMissingLevels moves down the tree a level towards the leaves until the node exists. This
+// must terminate successfully as we will eventually reach the leaves, which are always written
+// and are at level 0. Missing nodes are intermediate nodes with one child, hence their value
+// is the same as the node lower down the tree as there is nothing to hash it with.
+func skipMissingLevels(snapshot, lastNode int64, level int, node int64) (int, int64) {
```

Comment:
Maybe we should create an issue to sweep through the codebase to make it consistent?